### PR TITLE
[CELEBORN-1116] Read authentication configs from `HADOOP_CONF_DIR`

### PR DESCRIPTION
**Pull request by FMX**
### What changes were proposed in this pull request?
1. Make Celeborn read configs from HADOOP_COND_DIR.
2. Remove unnecessary Kerberos configs.

### Why are the changes needed?
To support HDFS with Kerberos.

### Does this PR introduce _any_ user-facing change?
NO.

### How was this patch tested?
GA and cluster.

### CODE REVIEW

File: common/src/main/scala/org/apache/celeborn/common/util/CelebornHadoopUtils.scala [hunk 1]
```diff
@@ -80,7 +80,7 @@ object CelebornHadoopUtils extends Logging {
     // If we are accessing HDFS and it has Kerberos enabled, we have to login
     // from a keytab file so that we can access HDFS beyond the kerberos ticket expiration.
     UserGroupInformation.setConfiguration(hadoopConf)
-    if (conf.hdfsStorageKerberosEnabled) {
+    if ("kerberos" == hadoopConf.get("hadoop.security.authentication")) {
```
Comment (pan3793) [inline common/src/main/scala/org/apache/celeborn/common/util/CelebornHadoopUtils.scala:83]: the value is case insensitive

**Comment by pan3793** [comment]
I missed the review in the previous PR.

keytab is enforced in the current implementation for Kerberos authentication, it's a common case for long-running services, but it's also possible to use TGT cache for Kerberos authentication (I know in some companies, TGT cache refresh is handled by the infra team in container/mechanism init phase, so that the application don't need to handle that and don't need to manage and periodically renew the keytab)

For example, `spark-submit` allows to use `--principal` and `--keytab`, but also allows to use TGT cache with `--proxy-user` to access the kerberized services.

We'd better document such limitations in the log message and docs clearly

**Comment by FMX** [comment]
Yes, as you have said spark supports using TGT cache. Celeborn does not support that yet. If using celeborn with kerberized HDFS, just setting the principle and keytab could be enough.

**Comment by pan3793** [comment]
it's quite easy to support auth from tgt cache

val kerberosEnabled = "KERBEROS".equalsIgnoreCase(hadoopConf.get("hadoop.security.authentication"))
if (kerberosEnabled) {
  val principalOpt = ...
  val keytabOpt = ...
  (principalOpt, keytabOpt) match {
    case (Some(principal), Some(keytab)) =>
      logInfo("Kerberos is enabled, attempt to login with principal: $principal and keytab: $keytab")
      UserGroupInformation.loginUserFromKeytab(principal, keytab)
    case _ =>
      logInfo("Kerberos is enabled, but without principal and keytab supplied, assuming keytab is managed externally")
      UserGroupInformation.getCurrentUser()
  }
}

**Comment by FMX** [comment]
I'll try it.

**Comment by FMX** [comment]
Thanks @pan3793 , you are right. I've test this PR on a cluster and it works well.

**Comment by pan3793** [comment]
@FMX thanks for the confirmation, have you tested the negative cases? e.g. access kerberized HDFS without keytab and TGT

**Comment by FMX** [comment]
I'll test the negative scenario in a moment.

**Comment by liujiayi771** [comment]
Not all users' Celeborn and HDFS are on the same cluster. If they are on different clusters, Celeborn may not have HADOOP_CONF_DIR. In this case, you can directly set the Hadoop parameters.

**Comment by pan3793** [comment]
@liujiayi771 what blocks you creating a folder(e.g. `/opt/hadoop/conf`) and put the Hadoop client configurations under that, then set `export HADOOP_CONF_DIR=/opt/hadoop/conf` in `celeborn-env.sh`?

**Comment by pan3793** [comment]
In practice, we usually add `HADOOP_CONF_DIR` into classpath so that Hadoop client could pick it up from classpath

File: common/src/main/scala/org/apache/celeborn/common/util/CelebornHadoopUtils.scala [hunk 2]
```diff
@@ -77,21 +77,22 @@ object CelebornHadoopUtils extends Logging {
...
-      if (!new File(keytab).exists()) {
-        throw new CelebornException(s"Keytab file: ${keytab} does not exist")
-      } else {
-        logInfo("Attempting to login to Kerberos " +
-          s"using principal: ${principal} and keytab: ${keytab}")
-        UserGroupInformation.loginUserFromKeytab(principal, keytab)
+    if ("kerberos".equals(hadoopConf.get("hadoop.security.authentication").toLowerCase)) {
+      val principalOpt = conf.hdfsStorageKerberosPrincipal
+      val keytabOpt = conf.hdfsStorageKerberosKeytab
+      (principalOpt, keytabOpt) match {
```
Comment (pan3793) [inline common/src/main/scala/org/apache/celeborn/common/util/CelebornHadoopUtils.scala:84]:
```suggestion
      (conf.hdfsStorageKerberosPrincipal, conf.hdfsStorageKerberosKeytab) match {
```

File: docs/deploy.md [hunk 3]
```diff
@@ -58,6 +58,10 @@ celeborn.rpc.askTimeout 240s
 celeborn.worker.flusher.hdfs.buffer.size 4m
 celeborn.storage.hdfs.dir hdfs://<namenode>/celeborn
 celeborn.worker.replicate.fastFail.duration 240s
+# If you HDFS is enabled with Kerberos. 
+# You will need to set the following configs or use kinit to get valid TGT.
```
Comment (pan3793) [inline docs/deploy.md:62]:
```suggestion
# Either principal/keytab or valid TGT cache is required to access kerberized HDFS
```

**Comment by liujiayi771** [comment]
I didn't mention which approach is better, I simply suggested providing explanations for both methods.

**Comment by pan3793** [comment]
@liujiayi771 Now, the behavior is exactly same with the Hadoop behavior, and the feature of hadoop configuration override is mentioned at https://celeborn.apache.org/docs/latest/migration/#upgrading-from-02-to-03

> Since 0.3.0, Celeborn supports overriding Hadoop configuration(core-site.xml, hdfs-site.xml, etc.) from Celeborn configuration with the additional prefix `celeborn.hadoop.`. On Spark client side, user should set Hadoop configuration like `spark.celeborn.hadoop.foo=bar`, note that `spark.hadoop.foo=bar` does not take effect; on Flink client and Celeborn Master/Worker side, user should set like `celeborn.hadoop.foo=bar`.

Also, we can emphasize that at https://celeborn.apache.org/docs/latest/configuration/ page too

**Review by pan3793 — APPROVED** [review]
LGTM

**Comment by FMX** [comment]
The negative case result looks OK to me. I believe that if users are using HDFS with Kerbero then they should know something about Kerberos.

> 23/11/09 10:14:52,256 WARN [celeborn-dispatcher-5] Client: Exception encountered while connecting to the server
> org.apache.hadoop.security.AccessControlException: Client cannot authenticate via:[TOKEN, KERBEROS]
> 	at org.apache.hadoop.security.SaslRpcClient.selectSaslClient(SaslRpcClient.java:173)

File: common/src/main/scala/org/apache/celeborn/common/util/CelebornHadoopUtils.scala [hunk 4]
```diff
@@ -77,21 +77,20 @@ object CelebornHadoopUtils extends Logging {
...
-        throw new CelebornException(s"Keytab file: ${keytab} does not exist")
-      } else {
-        logInfo("Attempting to login to Kerberos " +
-          s"using principal: ${principal} and keytab: ${keytab}")
-        UserGroupInformation.loginUserFromKeytab(principal, keytab)
+    if ("kerberos".equals(hadoopConf.get("hadoop.security.authentication").toLowerCase)) {
+      (conf.hdfsStorageKerberosPrincipal, conf.hdfsStorageKerberosKeytab) match {
+        case (Some(principal), Some(keytab)) =>
+          logInfo("Attempting to login to Kerberos " +
+            s"using principal: ${principal} and keytab: ${keytab}")
```
Comment (cfmcgrady) [inline common/src/main/scala/org/apache/celeborn/common/util/CelebornHadoopUtils.scala:85]:
```suggestion
          logInfo(
            s"Attempting to login to Kerberos using principal: $principal and keytab: $keytab")
```

File: common/src/main/scala/org/apache/celeborn/common/util/CelebornHadoopUtils.scala [hunk 5]
```diff
@@ -77,21 +77,20 @@ object CelebornHadoopUtils extends Logging {
...
+          if (!new File(keytab).exists()) {
+            throw new CelebornException(s"Keytab file: ${keytab} does not exist")
```
Comment (cfmcgrady) [inline common/src/main/scala/org/apache/celeborn/common/util/CelebornHadoopUtils.scala:87]:
```suggestion
            throw new CelebornException(s"Keytab file: $keytab does not exist")
```

**Review by cfmcgrady — APPROVED** [review]

**Pull request closed without merging**